File: 2021/day05-01.js
const fs = require('fs');

const input_file = process.argv.slice(2,3)[0];

let inputs = fs.readFileSync(input_file, 'utf8')
               .trim()
               .split('\n')
               .map( line => {
                 return line.split('->').map(coords => {
                   return coords.split(',').map(coord => {
                     return parseInt(coord)
                   })
                  })
                })

function values_between(a,b){
  let values = []
  // WTF sorting is done alphanumerically..?!
  let [min,max] = [a,b].sort( (c,d) => c -d )
  while(min<=max){
    values.push(min);
    min += 1
  }
  return values
}

let matrix = {}

function points(p1, p2){
  let points = [];

  if(p1[0] == p2[0]){ // Same X, so vertical line
    values_between(p1[1],p2[1]).forEach( y => points.push([p1[0], y]) )
  }
  else if(p1[1] == p2[1]){
    values_between(p1[0],p2[0]).forEach( x => points.push([x, p1[1]]) )
  }

  return points;
}

function draw_line(matrix, line_coords){
  let p1 = line_coords[0]
  let p2 = line_coords[1]

  start = Object.keys(matrix).length

  points(p1, p2).forEach( point => {
    if(point in matrix){
      matrix[point] += 1
    } else {
      matrix[point] = 1
    }
  })
}

inputs.forEach(coords => draw_line(matrix, coords))
intersections = Object.values(matrix)
                      .map(value => value > 1 ? 1 : 0)
                      .reduce( (prev, current) => prev + current )

console.log(intersections)

File: 2021/day05-01.rb
require 'byebug'

values = File.open(ARGV[0])
            .readlines
            .collect do |line|
              line.match(/(\d+),(\d+) -> (\d+),(\d+)/)
                  .to_a()[1..]
                  .collect { |v| v.to_i }
            end

matrix = Hash.new(0)

values.each do |x1, y1, x2, y2|
  next unless x1 == x2 || y1 == y2

  x1,x2 = [x1,x2].sort
  y1,y2 = [y1,y2].sort

  start = matrix.length

  if x1 == x2 # Vertical line
    (y1..y2).each { |y| matrix[ [x1, y] ] += 1 }
  elsif y1 == y2 # Horizontal
    (x1..x2).each { |x| matrix[ [x, y1] ] += 1 }
  end

  puts "#{ [x1,y1,x2,y2] } : #{matrix.length - start}"
end

intersections = matrix.values.select { |v| v > 1 }.count
puts(intersections)